### ISSUE TITLE
[feature]: testnet4 support

### ISSUE THREAD

**Issue by Impa10r** [lightningnetwork/lnd#8966]
**Is your feature request related to a problem? Please describe.**
I'd like to switch my app development to testnet4. LND gives error when I create or unlock a wallet:
`error creating wallet config: unable to create partial chain control: unknown network with genesis hash 00000000da84f2bafbbc53dee25a72ae507ff4914b867c565be350b0da8bf043`

**Describe the solution you'd like**
LND should start supporting testnet4. 

**Describe alternatives you've considered**
testnet3 is nearing EOL and is not compatible with liquidtestnet4

**Additional context**
btcd guys are working on this: https://github.com/btcsuite/btcd/issues/2187

**Comment by Roasbeef** [comment]
See the issue linked there, testnet4 hasn't really "officially" launched yet, as it isn't included yet in an final release of any Bitcoin full node daemon. It also includes new consensus rules that don't exist on mainnet, and will only exist on testnet4, so there may be some extra time tacked on to the end of development to make sure everything is compatible (eg: the BIP doesn't _yet_ have test vectors for these items).

**Comment by Impa10r** [comment]
Sure, I understand and appreciate all these issues. Testnet is for developing and testing out new things with no risk of funds loss. It does not have to be perfect. One can already build Bitcoin Core from [this PR]( https://github.com/bitcoin/bitcoin/pull/29775) to run with testnet4.

The reason we need LND to support it is because Liquid is nearing release of Elements Core v23.2.2 that will have discounted vSize for confidential transactions. PeerSwap users will greatly benefit from this, but the necessary development requires testnet4/liquidtestnet4.

**Comment by Roasbeef** [comment]
> One can already build Bitcoin Core from https://github.com/bitcoin/bitcoin/pull/29775 to run with testnet4.

Yep, I linked the PR in my comment above. My main point is that: that isn't released yet, so we can't consider any of the params or even the new consensus changes final. It also doesn't appear that it'll be a part of Bitcoind 28, as the feature freeze is in a week or so. If it's a part of Bitcoind 29, then that puts the actual release date sometime in 2025 IIUC. 

If you urgently need support, then depending on your available resources, you should be able to put together a temporary branch with the functionality. 

> Testnet is for developing and testing out new things with no risk of funds loss. It does not have to be perfect.

That `btcd` issues links to a partial version, but it doesn't have the new consensus rules, so it risks forking off of testnet4. 

Assuming the `btcd` PR is finalized in short order, the soonest it would land in `lnd` is ~November, as it would be coupled with lnd v0.19.

**Comment by Impa10r** [comment]
Thanks. Forking and fixing both lnd and btcd sounds too much work ).

**Comment by melvincarvalho** [comment]
I've been observing testnet4 for quite some time, and so far it is working very well.  Testnet3 which is on block 2.8 million is an interesting experiment but at times unusable.  There are also a lack of faucets.  I've collected resources on testnet4.  I hope it gets merged into 28, but bitcoin devs need something that just works for the next 5+ years or so.  I believe testnet4 will be the de-facto way of testing ideas on bitcoin.  Would be great if lnd joins the party sooner, rather than later.

https://github.com/testnet4/awesome-testnet4

**Comment by melvincarvalho** [comment]
There is also an active BIP discussion:  https://github.com/bitcoin/bips/pull/1601#issuecomment-2263490441

**Comment by melvincarvalho** [comment]
BIP-94 (Testnet4) is now merged

https://github.com/bitcoin/bips/blob/master/bip-0094.mediawiki

**Comment by Impa10r** [comment]
To whom it may concern, I forked LND/btcd/btcwallet with "naive" support for Testnet4. It has no DNS seeds, but I opened a channel to another node and paid an invoice. This is all I need for my purpose. I used lnd master, which shows 0.18.0-beta. Use bitcoin.testnet4=true in lnd.conf.

https://github.com/Impa10r/lnd

**Comment by melvincarvalho** [comment]
testnet4 is now merged into bitcoin master.  It is anticipated that testnet3 will be deprecated and removed in future versions.

https://github.com/bitcoin/bitcoin/pull/29775

**Comment by Roasbeef** [comment]
I think we'll wait a bit longer for the dust to settle re testnet4 (eg: potential param changes, reexamining difficult adjustment fixes). Until then, testnet3 still works and pretty much every production test environment for Bitcoin infrastructure still uses it.  

Testnet4 also isn't very useful for testing LN until there're a few hundred channels active on the new test network.

**Comment by Roasbeef** [comment]
Not completed, will keep this open to track integration.

**Comment by MegalithicBTC** [comment]
Just wanted to nudge this. I can see two classes of developers who might have differing perspectives on this issue.

1. Developers active for 4+ years, who have a big stash of testnet coins, and who might have a lot of existing scripts & code which is built around Testnet3.... and who CAN afford testing with testnet fees that look like this:
![image](https://github.com/user-attachments/assets/070074a8-e56f-4b5b-8d32-2372307fbc9c)
 
2. New developers, who currently CAN NOT work on development with LND, due to the impossibility of getting Testnet3 coins.

In our case, we're developing software which uses LND, and we'd like other users and developers to be able to work with it, and we essentially won't be able to proceed until LND supports Testnet4.

**Comment by MegalithicBTC** [comment]
One of the biggest challenges with Testnet3 is dealing with the block storms. When developing software for use with LND, trying to diagnose if an issue you are having is because of these block storms is ..... very challenging.... It must be really challenging or downright impossible right now if you are new developer trying to build with LND.....

![image](https://github.com/user-attachments/assets/388a55d8-01d0-4c59-ab29-a27712efb194)

**Comment by twofaktor** [comment]
https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-28.0.md#testnet4bip94-support

**Comment by MegalithicBTC** [comment]
I think I've identified that, during a testnet3 blockstorm, LND won't open channels, which I guess makes sense, because it's not fully synced to the chain. I get the error:

```
2024-10-05 17:18:14.797 [DBG] FNDG: Failing funding flow for pending_id=9e6a35528e4fba8b489b048fcdd50c584136a93f33437559f9f0ce835215313d: Synchronizing blockchain
```

**Comment by Impa10r** [comment]
Giving it a nudge as Testnet4 is now supported by Bitcoin Core 28.0. And testnet3 is now totally unusable with 4000 s/vb fees and no faucets.

**Comment by allenwillan** [comment]
I want to throw my hat in the request as well. Currently conducting research for my PhD and am struggling to utilize testnet3 and acquire enough coin to do anything.

Being able to begin populating the testnet4 LN would be hugely advantageous, considering that there exist functioning faucets for it.

**Comment by MegalithicBTC** [comment]
Adding my nudge here. LND is open-source software, but serious development work is now to some extent closed to new developers unless they are willing to risk real funds on mainnet or find a way to get a big stash of testnet coins. On our end we wasted many hours with testnet3 and have now given up and just lowered our expectations about what we can and cannot test.

**Comment by Roasbeef** [comment]
There are two active PRs working on adding testnet4 support to `btcd`:
  * https://github.com/btcsuite/btcd/pull/2274
  * https://github.com/btcsuite/btcd/pull/2275

So far none of them have added the new difficulty adjustment algorithm specific to testnet4, which will be where the bulk of testing+review will be spent, as there're no test vectors for it. 

We'll likely focus review on which ever adds it first.

**Comment by barrydeen** [comment]
Adding support for this feature, testnet3 is completely unusable now, signet extremely hard to get coins and very limited wallet support. Infrastructure providers need this to give reliable sandboxes to their clients. Happy to help with testing on this one.

**Comment by Impa10r** [comment]
> There are two active PRs working on adding testnet4 support to `btcd`:
> 
> * [feat: Add testnet4 chain support btcsuite/btcd#2274](https://github.com/btcsuite/btcd/pull/2274)
> * [Add testnet4 support btcsuite/btcd#2275](https://github.com/btcsuite/btcd/pull/2275)
> 
> So far none of them have added the new difficulty adjustment algorithm specific to testnet4, which will be where the bulk of testing+review will be spent, as there're no test vectors for it.
> 
> We'll likely focus review on which ever adds it first.

Hello. Btcd has [merged](https://github.com/btcsuite/btcd/pull/2339) support for testnet4

**Comment by twofaktor** [comment]
> Hello. Btcd has [merged](https://github.com/btcsuite/btcd/pull/2339) support for testnet4

Does this mean everything is finally ready to start working on LND support for TestNet4? It's taking a long time; it's difficult to test things in Lightning without a test network...

**Comment by guggero** [comment]
> It's taking a long time;

You're welcome to test and review the two PRs that are required to get this into `lnd`:
 - https://github.com/btcsuite/btcwallet/pull/994
 - https://github.com/lightninglabs/neutrino/pull/311

**Comment by guggero** [comment]
> don't know how to test them in isolation from lnd 😕

Here you go: https://github.com/lightningnetwork/lnd/pull/9620

**Comment by Roasbeef** [comment]
> It's taking a long time; it's difficult to test things in Lightning without a test network...

Testnet3 is still alive and active: https://mempool.space/testnet. 

Testnet4 is a thing as well, but seems to not be mining any transactions lately: https://mempool.space/testnet4
